src/lib/alarmMgr/AlarmManager.cpp

```diff
@@ -0,0 +1,240 @@
+*
+* AlarmManager::AlarmManager - 
+*/
+AlarmManager::AlarmManager():
+dbOk(true),
+notificationErrorLogInterval(0),
+badInputLogInterval(0)
+{
+}
+
+
+
+/* ****************************************************************************
+*
+* AlarmManager::AlarmManager - 
+*/
+AlarmManager::AlarmManager(int _notificationErrorLogInterval, int _badInputLogInterval):
+dbOk(true),
+notificationErrorLogInterval(_notificationErrorLogInterval),
+badInputLogInterval(_badInputLogInterval)
+{
+}  
+
+
+
+/* ****************************************************************************
+*
+* AlarmManager::notificationErrorLogIntervalSet - 
+*/
+void AlarmManager::notificationErrorLogIntervalSet(int _notificationErrorLogInterval)
+{
+  notificationErrorLogInterval = _notificationErrorLogInterval;
+}
+
+
+
+/* ****************************************************************************
+*
+* AlarmManager::badInputLogIntervalSet - 
+*/
+void AlarmManager::badInputLogIntervalSet(int _badInputLogInterval)
+{
+  badInputLogInterval = _badInputLogInterval;
+}
+
+
+
+/* ****************************************************************************
+*
+* AlarmManager::dbError - 
+*
+* Returns false if no action is taken/necessary, otherwise, true is returned.
+*/
+bool AlarmManager::dbError(const std::string& details)
+{
+  if (dbOk == false)
+  {
+    return false;
+  }
+
+  LM_E(("DB in an erroneous state (%s)", details.c_str()));
+  dbOk = false;
+  return true;
+}
+
+
+
+/* ****************************************************************************
+*
+* AlarmManager::dbErrorReset - 
+*
+* Returns false if no action is taken/necessary, otherwise, true is returned.
+*/
+bool AlarmManager::dbErrorReset(void)
+{
+  if (dbOk == true)
+  {
+    return false;
+  }
+
+  ++dbErrors;
+  LM_E(("Releasing Database alarms"));
+  dbOk = true;
+  return true;
+}
+
+
+
+/* ****************************************************************************
+*
+* AlarmManager::notificationError - 
+*
+* Returns false if no action is taken/necessary, otherwise, true is returned.
+*
+* NOTE
+* The number of notificationErrors per url is maintained in the map.
+* Right now that info is not used, but might be in the future.
+* To do so, we'd need another counter as well, to not forget the accumulated 
+* number each time the notificationErrors are reset.
+*/
+bool AlarmManager::notificationError(const std::string& url, const std::string& details)
+{
+  std::map<std::string, int>::iterator iter = notificationV.find(url);
+
+  ++notificationErrors;
+
+  if (iter != notificationV.end())  // Already exists - add to the 'url-specific' counter
+  {
+    iter->second += 1;
+  }
+  else
+  {
+    notificationV[url] = 1;
+  }
+
+  if (notificationErrorLogInterval != 0)
+  {
+    if ((notificationErrors % notificationErrorLogInterval) == 1)
+    {
+      LM_W(("Notification Error [%d] for %s: %s", notificationErrors, url.c_str(), details.c_str()));
+    }
+  }
+
+  return true;
```

Comment:
If the return value is used to mark transition, the right behaviour woult be that if the alarm (for that particular url) already exist, you have to return false, true otherwise. Currently, it is an unconditional true.